Language InterOp Blog based on cppyy and Cling paper from 2023

QuillPusher opened the pull request:
- High Level Blog Post to introduce to **Language Interoperability**, based on the following Paper:

- "**Efficient and Accurate Automatic Python Bindings with cppyy & Cling**" ([https://arxiv.org/pdf/2304.02712.pdf](https://arxiv.org/pdf/2304.02712.pdf))

- **Date** of the Blog is kept as the Date that the paper was published on. Later we can share this post on LinkedIn etc, using a  #tbt (ThrowBack Thursday) or Feature Highlight type post

- **tested** on localhost:

![Blog_sample3_cppyy](https://github.com/compiler-research/compiler-research.github.io/assets/130300172/791f9f27-c44e-48ff-970f-4d5c97336269)

QuillPusher commented:
Hello @maximusron 
Vassil suggested that you'd be an excellent candidate to review/contribute to this blog post 🚀 
Please have a look 😎

davidlange6 commented:
some meta comments
 - lets keep this positive. Instead of shortcomings, we should talk about the new functionality provided by our work. Its description is missing. Probably section 2 of the paper is a good source of information.
 - The benchmarks need some more description. In particular, in the blog article the speedups up pitiful. However, they are actually really awesome because its in comparison to a numpy-enabled operation. The text in the article just above the same table is useful
  - A conclusion is probably helpful.

QuillPusher commented:
@maximusron I've added you as a collaborator to my fork of the website (branch: [interop_blog_post](https://github.com/QuillPusher/compiler-research.github.io/tree/interop_blog_post)) in case you would like to directly make edits to the post. Any feedback would help

vgvassilev commented:
@maximusron ping.

maximusron commented:
@QuillPusher I have made some edits, can you send me the invite again?

QuillPusher commented:
> @QuillPusher I have made some edits, can you send me the invite again?

@maximusron Invite sent 👍 
let me know if it works

vgvassilev commented on _posts/2023-04-05-language-interoperability-using-cppyy-and-cling.md:
```diff
@@ -0,0 +1,174 @@
+---
+title: "The Next Step in Language Interoperability using Cling and cppyy"
```
```suggestion
title: "The Next Step in Language Interoperability using compiler-as-a-service and cppyy"
```

vgvassilev commented on _posts/2023-04-05-language-interoperability-using-cppyy-and-cling.md:
```diff
@@ -0,0 +1,174 @@
+layout: post
+excerpt: "Scientific software in constantly being challenged by enthusiasts trying to
+test the boundaries of programming languages, in search of better performance
+and simpler workflows. One such breakthrough was achieved using Cling, a C++
+interpreter, that has presented new possibilities with an incremental
+compilation infrastructure that is available at runtime."
```
```suggestion
and simpler workflows. Interactive C++ interpreters such as Cling and ClangRepl
presented new possibilities with an incremental compilation infrastructure that
 is available at runtime."
```

vgvassilev commented on _posts/2023-04-05-language-interoperability-using-cppyy-and-cling.md:
```diff
@@ -0,0 +1,174 @@
+{% endcapture %}
+
+Scientific software in constantly being challenged by enthusiasts trying to
+test the boundaries of programming languages, in search of better performance
+and simpler workflows. One such breakthrough was achieved using Cling, a C++
+interpreter, that has presented new possibilities with an incremental
+compilation infrastructure that is available at runtime.
```
Replace the this as suggested above.

vgvassilev commented on _posts/2023-04-05-language-interoperability-using-cppyy-and-cling.md:
```diff
@@ -0,0 +1,174 @@
+performance capabilities. But the interactive nature of Python and the gentler
+learning curve led to higher adoption rates elsewhere, and as a result, it saw
+exponential advancements in capabilities that were ideal for data science
+research. However, this does not discredit the usefulness of C++ in data
+science. A lot of research infrastructure is still rooted in C++ and benefits
+from some of its unique features (e.g., access to accelerators in
+heterogeneous computing environments). 
```
```suggestion
research. Python shines when steering infrastructure written in high
performance language such as C or C++. However, it is challenging to
write the glue layers between both languages for every package available
in the C++ ecosystem.
```

vgvassilev commented on _posts/2023-04-05-language-interoperability-using-cppyy-and-cling.md:
```diff
@@ -0,0 +1,174 @@
+while targeting either the CPU or the GPU, and providing interfaces to use the
+JITed closures from low-level libraries. Numba helps lower Python to machine
+code level and minimizes costly language crossings. In order to provide the
+missing links for this research, Numba was also integrated with cppyy (an
+automatic runtime bindings generator).
```
```suggestion
missing links for this research, Numba was also integrated with the cppyy project -- an
automatic runtime bindings generator based using interactive C++ to connect to the Python runtime.
```

QuillPusher commented:
Thanks @maximusron for the changes.

@vgvassilev requested changes have been made. Please see if this looks good overall
CC: @davidlange6

QuillPusher commented:
Hello @davidlange6 
Did you get a chance to have a look at this? It is in much better shape than before 😅

davidlange6 commented:
i did, but unfortunately the text is now very circular and difficult to understand what was accomplished or enabled. I do not yet have a clear idea of how to progress with it. But I have three suggestions

- add a simple but real example of code that is now supported that wasn't before
- decide if the article is about adding numba support to cppyy or adding cppyy support to numba
- avoid meaningless statements like "advanced C++ features" in favor of specific ones like "you can now call a c++ function in python loops that numba can interpret"

QuillPusher commented:
Thanks @davidlange6 , I'll look at this with fresh eyes to review the flow

QuillPusher commented:
@maximusron , can you spot a suitable code example:

> add a simple but real example of code that is now supported that wasn't before

P.s., @maximusron does the example presented in the paper qualify, or have you come across a better option?


![code](https://github.com/compiler-research/compiler-research.github.io/assets/130300172/b7e8bec7-ffb6-4be8-92d2-fc234721d60c)

maximusron commented:
@QuillPusher Yes this example should work, it is the more simple example

QuillPusher commented:
Thanks @aaronj0  😎 

Hello @davidlange6
Kindly review the following changes:

- Added the Code example suggested by Aaron
- Focused more on Achievements of the research
- Removed verbose sections about C++/Python
- Revamped the document structure with clear headings for better readability
